### PR TITLE
C#: Extract 'and' and 'or' patterns

### PR DESCRIPTION
**Pull request by tamasvajk**
The CFG for `and` and `or` patterns look like: [and.pdf](https://github.com/github/codeql/files/5873826/and.pdf), [or.pdf](https://github.com/github/codeql/files/5873827/or.pdf)

Missing

- [x] stats https://jenkins.internal.semmle.com/job/Stats/job/CSharp/32/
- [x] db upgrade folder
- [ ] diff job ~~https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/806/~~ ~~https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/809/~~ https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/813/
- [x] refresh change note date

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 1]
```diff
@@ -502,6 +507,29 @@ class NotPatternExpr extends UnaryPatternExpr, @not_pattern_expr {
...
+/** A binary pattern. For example, `1 or 2`. */
+class BinaryPatternExpr extends PatternExpr, @binary_pattern_expr {
+  /** Gets the left pattern. */
+  PatternExpr getLeftOperand() { result = this.getChild(0) }
+
+  /** Gets the right pattern. */
+  PatternExpr getRightOperand() { result = this.getChild(1) }
+}
+
+/** A binary or pattern. For example, `1 or 2`. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:522]: `or` -> `` `or` ``

File: csharp/ql/src/semmle/code/csharp/exprs/Expr.qll [hunk 2]
```diff
@@ -502,6 +507,29 @@ class NotPatternExpr extends UnaryPatternExpr, @not_pattern_expr {
...
+class OrPatternExpr extends BinaryPatternExpr, @or_pattern_expr {
+  override string toString() { result = "... or ..." }
+
+  override string getAPrimaryQlClass() { result = "OrPatternExpr" }
+}
+
+/** A binary and pattern. For example, `< 1 and > 2`. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:529]: `and` -> `` `and` ``

File: csharp/ql/src/semmle/code/csharp/controlflow/internal/ControlFlowGraphImpl.qll [hunk 3]
```diff
@@ -51,12 +51,21 @@ private import Splitting
...
 
 module ControlFlowTree {
   private class Range_ = @callable or @control_flow_element;
 
-  class Range extends Element, Range_ { }
+  class Range extends Element, Range_ {
+    Range() { this = getAChild*(any(CfgScope scope)) }
+  }
+
+  private Element getAChild(Element p) {
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/controlflow/internal/ControlFlowGraphImpl.qll:65]: Use this in `getAChildInScope`.

**Review by hvitved — CHANGES_REQUESTED** [review]
Looks great!

**Review by hvitved — DISMISSED** [review]

**Comment by tamasvajk** [comment]
The new pattern matching completion changes result in a lot of noise or false positives in the new constant condition check.

**Comment by hvitved** [comment]
> The new pattern matching completion changes result in a lot of noise or false positives in the new constant condition check.

That should be fixed by the CFG changes I was working on, so happy to merge this now.

**Review by hvitved — APPROVED** [review]

**Pull request merged**